### ISSUE TITLE
Build paths in installed .cmake files

### ISSUE THREAD

**Issue by rossburton** [libical/libical#532]
Two of the installed `.cmake` files have build paths in:

`/usr/lib/cmake/LibIcal/IcalGlibSrcGenerator.cmake`:
```
INTERFACE_COMPILE_OPTIONS "-I/yocto/ross/build/tmp/work/armv8a-poky-linux/libical/3.0.12-r0/recipe-sysroot/usr/include/glib-2.0;-I/yocto/ross/build/tmp/work/armv8a-poky-linux/libical/3.0.12-r0/recipe-sysroot/usr/lib/glib-2.0/include;-I/yocto/ross/build/tmp/work/armv8a-poky-linux/libical/3.0.12-r0/recipe-sysroot/usr/include/libxml2;-DG_LOG_DOMAIN=\"src-generator\""
```
`/usr/lib/cmake/LibIcal/LibIcalTargets.cmake`:
```
INTERFACE_LINK_LIBRARIES "/yocto/ross/build/tmp/work/armv8a-poky-linux/libical/3.0.12-r0/recipe-sysroot/usr/lib/libicuuc.so;/yocto/ross/build/tmp/work/armv8a-poky-linux/libical/3.0.12-r0/recipe-sysroot/usr/lib/libicui18n.so"
```

**Comment by winterz** [comment]
@lopsided98 do you any thoughts about this?

**Comment by lopsided98** [comment]
Sorry for the very late reply, but the first one can probably be fixed by changing the `target_compile_options` call from `PUBLIC` to `PRIVATE`:

https://github.com/libical/libical/blob/88e9115909e8fbaaaa2e1b952ad3f66b0e9c0c7c/src/libical-glib/CMakeLists.txt#L18-L19

`target_compile_options` already used `PUBLIC` before my changes, but the target wasn't exported so the compile options didn't end up in the installed cmake files. The `target_link_libraries` call should probably use `PRIVATE` as well.

I think the second one is unrelated to my changes, but caused by a similar `PUBLIC` or unqualified usage of `target_link_libraries`, probably this one: https://github.com/libical/libical/blob/88e9115909e8fbaaaa2e1b952ad3f66b0e9c0c7c/src/libical/CMakeLists.txt#L297C39. More work is needed to determine if this can be made `PRIVATE` or whether it is already correct.

On the other hand, even if these specific cases can be made private, there are certainly cases where public dependencies are necessary, and it seems like this would always break bitbake/Yocto. I found [this post describing a similar problem](https://discourse.cmake.org/t/problematic-absolute-paths-in-cmake-file-created-by-install-export/2246) that suggests that the fix is to use imported targets that use library/include paths relative to a sysroot variable. There are a lot of CMake projects out there that don't use imported targets though, and I can't imagine that bitbake doesn't support them.

**Comment by mcrha** [comment]
In an ideal world, the `ical-glib-src-generator` is not installed at all. It's an internal/private tool to convert API .xml files into a code, which is then built in the libical-glib bindings. It has no use on the user machine (for example Fedora deletes it). I understood, the install is needed to be able to cross-compile the sources, unfortunately.

**Comment by rossburton** [comment]
Yeah, CMake doesn't really support "build this binary for the native machine so I can run it during the build" nicely. You can do tricks with externalproject but I wouldn't be surprised if it was easier to port the build to Meson...

**Comment by mcrha** [comment]
I do not have much opinion on the change (option to (not) install the generator). It works as it is for quite some time now. The distros can handle it in the way they prefer.

**Comment by winterz** [comment]
https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html#variable:CMAKE_CROSSCOMPILING

what about using the CMAKE_CROSSCOMPILING cmake variable.
if CMAKE_CROSSCOMPILING then don't do some stuff.  although I'm not sure what needs to not be done.

**Comment by rossburton** [comment]
Doesn't quite work: when cross-compiling you need an existing generator, when not cross-compiling you build your own and may or may not need to install it.

**Comment by rossburton** [comment]
Verified with 4.0.4. Without any of our tweaks in a cross-compiled environment with sysroots:

```
ERROR: libical-4.0.4-r0 do_package_qa: QA Issue: File /usr/lib/cmake/LibIcal/IcalGlibSrcGenerator.cmake in package libical-dev contains reference to TMPDIR [buildpaths]
ERROR: libical-4.0.4-r0 do_package_qa: QA Issue: File /usr/lib/cmake/LibIcal/LibIcalTargets.cmake in package libical-dev contains reference to TMPDIR [buildpaths]
```

Specifically:
```
usr/lib/cmake/LibIcal/IcalGlibSrcGenerator.cmake:  INTERFACE_COMPILE_OPTIONS "-I/work/ross/build/tmp/work/core2-32-poky-linux/libical/4.0.4/recipe-sysroot/usr/include/glib-2.0;-I/work/ross/build/tmp/work/core2-32-poky-linux/libical/4.0.4/recipe-sysroot/usr/lib/glib-2.0/include;-I/work/ross/build/tmp/work/core2-32-poky-linux/libical/4.0.4/recipe-sysroot/usr/include;-DG_LOG_DOMAIN=\"src-generator\""
usr/lib/cmake/LibIcal/LibIcalTargets.cmake:  INTERFACE_LINK_LIBRARIES "/work/ross/build/tmp/work/core2-32-poky-linux/libical/4.0.4/recipe-sysroot/usr/lib/libicui18n.so;/work/ross/build/tmp/work/core2-32-poky-linux/libical/4.0.4/recipe-sysroot/usr/lib/libicuuc.so;/work/ross/build/tmp/work/core2-32-poky-linux/libical/4.0.4/recipe-sysroot/usr/lib/libicudata.so"
```

The first is solved by #1010.

**Comment by rossburton** [comment]
The latter is "fixed" by changing `src/libical/CMakeLists.txt`:
```
-target_link_libraries(ical ${ICU_LIBRARIES})
+target_link_libraries(ical PRIVATE ${ICU_LIBRARIES})
```
At least it's not exposing the ICU linkage in the .cmake file, which shouldn't be required by external users as the library already links.

As I'm sure I've said, I'm no cmake-expert but I _think_ this is the correct behaviour.

**Comment by rossburton** [comment]
I just pushed https://github.com/libical/libical/pull/1366 as a "throw something as a PR and see what happens" test. That minimal change removes the build path and doesn't appear to break the builds for me at least.
